### PR TITLE
Sycl implementation for objective functions

### PR DESCRIPTION
**Pull request by razdoburdin**
In continuation of https://github.com/dmlc/xgboost/pull/9800
Replaces of https://github.com/dmlc/xgboost/pull/7732

This PR adds SYCL implementation and corresponding tests for `multi:softmax`, `multi:softprob`, `reg:squarederror`, `reg:squaredlogerror`, `reg:logistic`, `binary:logistic`, `binary:logitraw` objectives.

The dispatching is organized as follows. In case of execution device is set to `sycl`, the original name of objective function is replaced by `original_name_sycl`. In case of requited function hasn't sycl implementation, the original name is used.

### CODE REVIEW

File: plugin/sycl/objective/multiclass_obj.cc [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+ * \tparam Iterator The type of the iterator.
+ */
+template<typename Iterator>
+inline Iterator FindMaxIndex(Iterator begin, Iterator end) {
+  Iterator maxit = begin;
+  for (Iterator it = begin; it != end; ++it) {
+    if (*it > *maxit) maxit = it;
+  }
+  return maxit;
+}
```
Comment (trivialfis) [inline plugin/sycl/objective/multiclass_obj.cc:73]: Any chance we can use the one in `common/math.h`?

File: plugin/sycl/objective/multiclass_obj.cc [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+struct SoftmaxMultiClassParam : public XGBoostParameter<SoftmaxMultiClassParam> {
+  int num_class;
+  // declare parameters
+  DMLC_DECLARE_PARAMETER(SoftmaxMultiClassParam) {
+    DMLC_DECLARE_FIELD(num_class).set_lower_bound(1)
+        .describe("Number of output class in the multi-class classification.");
+  }
+};
```
Comment (trivialfis) [inline plugin/sycl/objective/multiclass_obj.cc:82]: Maybe create a header instead of copying?

File: plugin/sycl/objective/regression_loss.h [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  // duplication is necessary, as __device__ specifier
+  // cannot be made conditional on template parameter
+  static bst_float PredTransform(bst_float x) { return Sigmoid(x); }
+  static bool CheckLabel(bst_float x) { return x >= 0.0f && x <= 1.0f; }
+  static bst_float FirstOrderGradient(bst_float predt, bst_float label) {
+    return predt - label;
+  }
+  static bst_float SecondOrderGradient(bst_float predt, bst_float label) {
+    const bst_float eps = 1e-16f;
+    return std::max(predt * (1.0f - predt), eps);
```
Comment (trivialfis) [inline plugin/sycl/objective/regression_loss.h:85]: So .. is this `sycl::max` or `std::max`? If the compiler can handle `std::max` then we can avoid lots of copy and paste.

File: plugin/sycl/objective/regression_loss.h [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    return 1.0f;
+  }
+  static bst_float ProbToMargin(bst_float base_score) { return base_score; }
+  static const char* LabelErrorMsg() { return ""; }
+  static const char* DefaultEvalMetric() { return "rmse"; }
+
+  static const char* Name() { return "reg:squarederror_sycl"; }
+
+  static ObjInfo Info() { return {ObjInfo::kRegression, true, false}; }
+};
```
Comment (trivialfis) [inline plugin/sycl/objective/regression_loss.h:44]: Looks like the same as the original one.

File: plugin/sycl/objective/regression_loss.h [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  static T FirstOrderGradient(T predt, T label) { return predt - label; }
+  template <typename T>
+  static T SecondOrderGradient(T predt, T label) {
+    const T eps = T(1e-16f);
+    return std::max(predt * (T(1.0f) - predt), eps);
+  }
+  static bst_float ProbToMargin(bst_float base_score) {
+    CHECK(base_score > 0.0f && base_score < 1.0f)
+        << "base_score must be in (0,1) for logistic loss, got: " << base_score;
+    return -logf(1.0f / base_score - 1.0f);
```
Comment (trivialfis) [inline plugin/sycl/objective/regression_loss.h:99]: > // TODO(razdoburdin): SYCL does not fully support std math inside offloaded kernels

Would be great if we could avoid copying when the function is supported.

File: src/learner.cc [hunk 6]
```diff
@@ -528,6 +528,9 @@ class LearnerConfiguration : public Learner {
     auto const& objective_fn = learner_parameters.at("objective");
     if (!obj_) {
       CHECK_EQ(get<String const>(objective_fn["name"]), tparam_.objective);
+      if (ctx_.IsSycl()) {
+        tparam_.objective = ObjFunction::GetSyclImplementationName(tparam_.objective);
+      }
```
Comment (trivialfis) [inline src/learner.cc:533]: Is it possible that the dispatching happens inside `ObjFunction::Create`? Also, what happens if a user saves the model, loads it back on a different machine that doesn't have a sycl device?

**Review by trivialfis — COMMENTED** [review]
Hi, thank you for the work on objective functions. Some initial comments for discussion. It would be great if we could reduce the amount of copied code.

> Find a better way to dispatch names of SYCL kernels with various

If you have ideas on how to refactor the code, please share. ;-)

File: src/learner.cc (hunk 6 above)
Comment (razdoburdin) [reply src/learner.cc]: It is a good idea. In this case we wouldn't have problems with saved models.

File: plugin/sycl/objective/multiclass_obj.cc (hunk 1 above)
Comment (razdoburdin) [reply plugin/sycl/objective/multiclass_obj.cc]: done

**Review by trivialfis — APPROVED** [review]
